### PR TITLE
Implement Anti-Radiation Ammunition (ARAD)

### PR DESCRIPTION
**Pull request by HammerGS**
Implement ARAD (Anti-Radiation) Missiles for LRM/SRM/MML

  ## Summary

  Implements ARAD (Anti-Radiation) missile functionality for LRM, SRM, and MML weapon systems per Tactical
  Operations: Advanced Units & Equipment, p.180.

  ARAD missiles receive bonuses against targets with active electronics and penalties against targets without
  electronics. ECM can block the bonus unless the target is Narc-tagged.

  ## Changes

  ### Core Implementation

  **Equipment Detection System:**
  - `ARADEquipmentDetector.java` - Centralized electronics detection utility
    - Detects: Active Probe, Artemis IV/V, Blue Shield, C3 (all types), Communications Equipment ≥3.5 tons, ECM,
  TAG, Ghost Targets, Narc/iNarc pods
    - Handles Active Stealth Armor blocking (blocks internal systems, NOT external Narc)
    - Validates equipment state (not destroyed/missing/breached)
    - Friend-or-foe Narc pod detection

  **LRM ARAD Handler:**
  - `LRMARADHandler.java` - Weapon handler for LRM ARAD missiles
    - Cluster modifiers: +1 (electronics), 0 (ECM blocked), -2 (no electronics)
    - Narc overrides ECM blocking per TO:AUE p.180
    - Report messages: 3363 (ECM blocked), 3364 (Narc override), 3368 (no electronics), 3369 (electronics detected)

  **SRM ARAD Handler:**
  - `SRMARADHandler.java` - Weapon handler for SRM ARAD missiles
    - Identical cluster logic to LRM
    - Full report message support

  **To-Hit Modifiers:**
  - `ComputeToHit.java` - Added ARAD to-hit calculation
    - -1 to-hit vs electronics (unless ECM blocked)
    - +2 to-hit vs no electronics
    - Narc override for ECM blocking
    - Non-entity target handling (buildings)

  **Weapon Routing:**
  - `LRMWeapon.java` - Added M_ARAD munition routing
  - `SRMWeapon.java` - Added M_ARAD munition routing
  - `MMLWeapon.java` - No changes needed (automatically routes to LRM/SRM handlers)

  **Localization:**
  - `messages.properties` - Added 4 ARAD report messages
    - 3363: ECM blocked bonus
    - 3364: Narc override
    - 3368: No electronics penalty
    - 3369: Electronics detected bonus

  ### Bug Fixes

  **Communications Equipment Tonnage Detection:**
  - Fixed bug where variable-sized Communications Equipment reported base tonnage (1.0) instead of actual mounted
  tonnage (7.0)
  - Changed from `equipment.getType().getTonnage(target)` to `equipment.getTonnage()`
  - Validated with Mobile HQ (7 tons) now correctly triggering ARAD bonus

  **Building To-Hit Modifier:**
  - Fixed missing +2 to-hit modifier when targeting buildings
  - Added non-entity target handling in ComputeToHit.java

  **JavaDoc Parameter Mismatch:**
  - Fixed `@param ae` → `@param weaponEntity` in ComputeToHit.java

  ### Unit Tests

  **ARADEquipmentDetectorTest.java** - 27 comprehensive tests (100% passing)
  - All equipment type detection (Active Probe, Artemis, Blue Shield, C3, ECM, Communications, TAG, Ghost Targets,
  Narc/iNarc)
  - Stealth Armor blocking and Narc override behavior
  - Destroyed/missing/breached equipment validation
  - Heavy communications tonnage thresholds (≥3.5 tons)
  - Multiple system combinations and edge cases

  **LRMARADHandlerTest.java** - 7 scenario tests
  - Electronics detection with/without ECM
  - Narc override mechanics
  - Stealth Armor blocking
  - TAG and Ghost Targets

  **SRMARADHandlerTest.java** - 7 scenario tests (identical coverage to LRM)

  ## Test Plan

  ### Manual Testing Completed ✅

  **LRM ARAD (8/8 scenarios):**
  - Normal bonus: -1 to-hit, +1 cluster
  - ECM blocking: -1 to-hit, 0 cluster
  - Narc override: -1 to-hit, +1 cluster (ECM ignored)
  - No electronics: +2 to-hit, -2 cluster
  - Building target: +2 to-hit, -2 cluster
  - Stealth Armor: Blocks internal, allows Narc
  - Communications Equipment: 7-ton equipment detected
  - Complex multi-unit scenarios validated

  **SRM ARAD (4/4 core scenarios):**
  - Normal bonus, ECM blocking, Narc override, no electronics
  - All results match LRM behavior ✅

  **MML ARAD:**
  - LRM mode routes to LRMARADHandler ✅
  - SRM mode routes to SRMARADHandler ✅
  - No implementation needed (automatic routing)

  ### Unit Testing ✅
  - ARADEquipmentDetectorTest: 27/27 passing
  - All equipment detection logic validated
  - Edge cases and combinations tested

  ## Rules Reference

  **Tactical Operations: Advanced Units & Equipment, p.180**
  - ARAD missiles detect active electronics
  - Bonuses: -1 to-hit, +1 cluster vs electronics
  - Penalties: +2 to-hit, -2 cluster vs no electronics
  - ECM blocks bonus unless target is Narc-tagged
  - Narc overrides ECM blocking (no additional Narc bonus stacking)

  **Forum Rulings (documented in code):**
  - Stealth blocking: https://battletech.com/forums/index.php?topic=78845.msg1866412#msg1866412
  - Powered-down equipment: https://battletech.com/forums/index.php?topic=31896.msg1369654#msg1369654
  - Communications threshold: https://battletech.com/forums/index.php?topic=17456.msg396440#msg396440
  - Narc/iNarc interaction: https://battletech.com/forums/index.php?topic=26824.msg609067#msg609067

  ## Files Modified

  **New Files:**
  - megamek/src/megamek/common/weapons/handlers/ARADEquipmentDetector.java
  - megamek/src/megamek/common/weapons/handlers/lrm/LRMARADHandler.java
  - megamek/src/megamek/common/weapons/handlers/srm/SRMARADHandler.java
  - megamek/unittests/megamek/common/weapons/handlers/ARADEquipmentDetectorTest.java
  - megamek/unittests/megamek/common/weapons/handlers/lrm/LRMARADHandlerTest.java
  - megamek/unittests/megamek/common/weapons/handlers/srm/SRMARADHandlerTest.java

  **Modified Files:**
  - megamek/src/megamek/common/actions/compute/ComputeToHit.java
  - megamek/src/megamek/common/weapons/lrms/LRMWeapon.java
  - megamek/src/megamek/common/weapons/srms/SRMWeapon.java
  - megamek/resources/megamek/common/messages.properties

  ## Code Quality

  - All logging at DEBUG level (production-ready)
  - Comprehensive JavaDoc documentation
  - Forum ruling references in code comments
  - Single Source of Truth (SSOT) - ARADEquipmentDetector centralizes all detection logic
  - DRY principle - LRM/SRM handlers share identical logic patterns
  - Extensive unit test coverage (27 equipment detection tests)
  
   ## Next Steps

  - QA validation of game balance
  - Player feedback on ARAD mechanics

### CODE REVIEW

File: docs/issues/features/arad-implementation.md
Comment (IllianiBird) [inline docs/issues/features/arad-implementation.md:1]: Make sure to remove these AI config files before you put this up for review.

File: megamek/mmconf/log4j2.xml [hunk 1]
```diff
@@ -1,5 +1,5 @@
 <?xml version="1.0" encoding="utf-8"?>
-<Configuration status="info">
```
Comment (IllianiBird) [inline megamek/mmconf/log4j2.xml:2]: Make sure this is reverted before you flip this PR to live

File: megamek/mmconf/defaultKeyBinds.xml [hunk 2]
```diff
@@ -43,49 +43,49 @@
     </KeyBind>
 
     <KeyBind>
-        <command>deployTurnUnit</command> <!-- Shift-T -->
```
Comment (IllianiBird) [inline megamek/mmconf/defaultKeyBinds.xml:46]: These changes appear unrelated to the intent of this PR and should either be rolled back or put into a separate PR

File: megamek/resources/megamek/client/messages_es.properties [hunk 3]
```diff
@@ -3361,6 +3361,8 @@ WeaponAttackAction.AccWeapon=arma precisa
 WeaponAttackAction.AcRapid=Modo de disparo r\u00E1pido AC
 WeaponAttackAction.AdjustedFire=fuego ajustado
 WeaponAttackAction.AmmoMod=modificador de impacto de munici\u00F3n
+WeaponAttackAction.AradElectronics=Anti-Radiation - electronics detected
+WeaponAttackAction.AradNoElectronics=Anti-Radiation - no electronics detected
```
Comment (IllianiBird) [inline megamek/resources/megamek/client/messages_es.properties]: I don't believe you need to duplicate these entries, if an entry is missing for the user's selected language it will fallback to English

File: megamek/src/megamek/common/actions/compute/ComputeToHit.java [hunk 4]
```diff
@@ -238,6 +240,14 @@ public static ToHitData toHitCalc(Game game, int attackerId, Targetable target,
               (munition.contains(AmmoType.Munitions.M_FOLLOW_THE_LEADER) &&
                     !ComputeECM.isAffectedByECM(ae, ae.getPosition(), target.getPosition()));
 
+        boolean bARAD = (ammoType != null) &&
```
Comment (IllianiBird) [inline megamek/src/megamek/common/actions/compute/ComputeToHit.java]: Appending the class type in the variable (the 'b' in bARAD) is old fashioned and not in keeping with the kind of code we're pushing for. Any new code needs to use descriptive variable names that state what the variable is about.

In this instance there is no easy way for me to tell what `bARAD` is actually for, beyond it being a boolean that is in some way related to ARAD

File: megamek/src/megamek/common/actions/compute/ComputeToHit.java [hunk 5]
```diff
@@ -238,6 +240,14 @@ public static ToHitData toHitCalc(Game game, int attackerId, Targetable target,
...
+              ((ammoType.getAmmoType() == AmmoType.AmmoTypeEnum.LRM) ||
+                    (ammoType.getAmmoType() == AmmoType.AmmoTypeEnum.LRM_IMP) ||
+                    (ammoType.getAmmoType() == AmmoType.AmmoTypeEnum.SRM) ||
+                    (ammoType.getAmmoType() == AmmoType.AmmoTypeEnum.SRM_IMP) ||
+                    (ammoType.getAmmoType() == AmmoType.AmmoTypeEnum.MML)) &&
```
Comment (IllianiBird) [inline megamek/src/megamek/common/actions/compute/ComputeToHit.java]: Generally, massive streams of several clauses in a conditional are not great. They make for hard reading and it's _very_ easy to make logical mistakes when handling them. Far better to break out the individual clauses into their own boolean

There's an example of how this case work later in my review

File: megamek/src/megamek/common/weapons/handlers/ARADEquipmentDetector.java [hunk 6]
```diff
@@ -0,0 +1,351 @@
...
+ * (at your option) any later version.
+ *
+ * MegaMek is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with MegaMek. If not, see <http://www.gnu.org/licenses/>.
+ */
```
Comment (IllianiBird) [inline megamek/src/megamek/common/weapons/handlers/ARADEquipmentDetector.java:31]: This legal statement doesn't look right. It's missing the Microsoft stuff. Best to double-check it.

File: megamek/src/megamek/common/weapons/handlers/ARADEquipmentDetector.java [hunk 7]
```diff
@@ -0,0 +1,351 @@
...
+ *
+ * Key Forum Rulings:
+ * - Stealth blocking: https://battletech.com/forums/index.php?topic=78845.msg1866412#msg1866412
+ * - Powered-down equipment: https://battletech.com/forums/index.php?topic=31896.msg1369654#msg1369654
+ * - Communications threshold: https://battletech.com/forums/index.php?topic=17456.msg396440#msg396440
+ * - Narc/iNarc interaction: https://battletech.com/forums/index.php?topic=26824.msg609067#msg609067
+ * - General exclusions: https://battletech.com/forums/index.php?topic=63179.msg1452217#msg1452217
+ *
+ * @author MegaMek Team
+ * @since 2025-01-16
```
Comment (IllianiBird) [inline megamek/src/megamek/common/weapons/handlers/ARADEquipmentDetector.java]: If you're including an author tag you want it to signpost back to you. I'm a big proponent of author tags, they're super useful to aid knowing who was the original author of a class or method. But if you put something generic it loses that usefulness. Far better to put 'Hammer' here, instead.

Regarding the 'since' tag, this is less cut and dry. I use the version the class was added in myself

File: megamek/src/megamek/common/weapons/handlers/ARADEquipmentDetector.java [hunk 8]
```diff
@@ -0,0 +1,351 @@
...
+        // Priority 2: Check all qualifying systems
+        return hasActiveProbe(target) ||
+               hasArtemis(target) ||
+               hasBlueShield(target) ||
+               hasC3(target) ||
+               hasHeavyComms(target) ||
+               hasECM(target) ||
+               hasActiveTAG(target) ||
+               hasGhostTargets(target) ||
+               isNarcTagged(target, friendlyTeam);
```
Comment (IllianiBird) [inline megamek/src/megamek/common/weapons/handlers/ARADEquipmentDetector.java:100]: Earlier in this review I mentioned that chained conditional clauses isn't great code. This is an example of where it can work. Here each clause is very clear in what it's checking against. I, as a developer, can easily intuit what clause handles what and it's unlikely mistakes in logic will be introduced if the code is ever changed

File: megamek/mmconf/log4j2.xml (hunk 1 above)
Comment (HammerGS) [reply megamek/mmconf/log4j2.xml:2]: Done

File: megamek/mmconf/defaultKeyBinds.xml (hunk 2 above)
Comment (HammerGS) [reply megamek/mmconf/defaultKeyBinds.xml:46]: Done

File: megamek/src/megamek/common/actions/compute/ComputeToHit.java (hunk 4 above)
Comment (HammerGS) [reply megamek/src/megamek/common/actions/compute/ComputeToHit.java]: Addressed

File: megamek/src/megamek/common/weapons/handlers/ARADEquipmentDetector.java (hunk 7 above)
Comment (HammerGS) [reply megamek/src/megamek/common/weapons/handlers/ARADEquipmentDetector.java]: Addressed

**Comment by HammerGS** [comment]
I've retested thing today, and Lanceman also tested. Marking this ready for review. 

image

File: megamek/src/megamek/common/actions/compute/ComputeToHit.java [hunk 9]
```diff
@@ -794,13 +815,14 @@ public static ToHitData toHitCalc(Game game, int attackerId, Targetable target,
...
      * @param bHeatSeeking  flag that indicates whether the attacker is using Heat Seeking missiles
+     * @param isAradAttack  flag that indicates whether the attacker is using ARAD missiles
      * @param isECMAffected flag that indicates whether the target is inside an ECM bubble
      * @param isINarcGuided flag that indicates whether the target is broadcasting an iNarc beacon
      */
     private static ToHitData compileAmmoToHitMods(Game game, Entity ae, Targetable target, int targetType,
           ToHitData toHit, WeaponType weaponType, Mounted<?> weapon, AmmoType ammoType,
           EnumSet<AmmoType.Munitions> munition, boolean bApollo, boolean bArtemisV, boolean bFTL, boolean bHeatSeeking,
-          boolean isECMAffected, boolean isINarcGuided) {
+          boolean isAradAttack, boolean isECMAffected, boolean isINarcGuided) {
```
Comment (Sleet01) [inline megamek/src/megamek/common/actions/compute/ComputeToHit.java:825]: @HammerGS not for this patch, but we should examine turning all of these booleans into an extensible Record so we can replace the bottom two lines with a single object.

That way we'll be able to add new switches without changing the function signatures.

Comment (HammerGS) [reply megamek/src/megamek/common/actions/compute/ComputeToHit.java:825]: I'll add that to the Prompt.
